controllers/nutanixmachine_controller.go

```diff
@@ -514,6 +521,33 @@ func (r *NutanixMachineReconciler) reconcileNormal(rctx *nctx.MachineContext) (r
 	return reconcile.Result{}, nil
 }
 
+// checkFailureDomainStatus checks and sets the NutanixMachine.status.failureDomain if necessary
+func (r *NutanixMachineReconciler) checkFailureDomainStatus(rctx *nctx.MachineContext) error {
+	if rctx.Machine.Spec.FailureDomain == nil || *rctx.Machine.Spec.FailureDomain == "" {
+		return nil
+	}
+
+	// Fetch the referent failure domain object
+	fdObj := &infrav1.NutanixFailureDomain{}
+	fdName := *rctx.Machine.Spec.FailureDomain
+	fdKey := client.ObjectKey{Name: fdName, Namespace: rctx.NutanixMachine.Namespace}
+	err := r.Get(rctx.Context, fdKey, fdObj)
+	if err != nil {
+		return fmt.Errorf("failed to fetch the referent failure domain object %q: %w", fdName, err)
+	}
+
+	// Validate the NutanixMachine machine spec is consistent with that in the failure domain spec
+	if !reflect.DeepEqual(rctx.NutanixMachine.Spec.Cluster, fdObj.Spec.PrismElementCluster) ||
+		!reflect.DeepEqual(rctx.NutanixMachine.Spec.Subnets, fdObj.Spec.Subnets) {
+		return fmt.Errorf("The nutanixmachine's spec cluster/subnets is not consistent with that in the failure domain %q", *rctx.Machine.Spec.FailureDomain)
```

Comment:
Did you consider comparing the fields individually, and reporting a detailed diff to the user? I feel that reporting "is not consistent" may be too little information.

Also, did you consider implementing this comparison in an `Equal` receiver of `NutanixResourceIdentifier`?

Reply:
changed as suggested.

Reply:
Thank you!